### ISSUE TITLE
Monotonic tag projection

### ISSUE THREAD

**Issue by EtienneAb3d** [modernmt/modernmt#556]
MMT tag projection can change the order of the sequence of tags in a sentence. This may cause a DTD-invalid result. This is especially damageable in full-automatic process. 

Here is an experimental code for a monotonic tag projection. Of course, it will produce lower quality on sentences where words are strongly reordered. But, it will ensure that the final document is always DTD-valid (thus won't crash in a process using MS Office, for example).

It would be nice if such a tag projection could be added as an option, either at the MMT starting time, or at the translation request time.
[TagProjector.java.zip](https://github.com/modernmt/modernmt/files/5108966/TagProjector.java.zip)

PS : I forgot to remove some commented lines from past versions. Of course, all commented lines of code are not pertinent.

**Comment by nicolabertoldi** [comment]
Hi @EtienneAb3d 

thank you for your feedback.

In ModernMT we paid strong attention that tag projection respect the consistency of tags found of the source text.
In other words, if the tags are  consistent  in the source, it should be consistent in the output as well.

Hence, I am surprised from your issue report, and we want to debug

To this purpose, could you please provide us:
- the original text with the tags
- the translation with tags
- the source input (tokenized)
- the generated output (tokenized)
- the word alignment
Last three items can be found in the node.log when the engine is started with option (``-v 2``)

**Comment by EtienneAb3d** [comment]
It can be demonstrated with any very simple cases. 

For example:
`Ceci est un <t1>chapeau</t1> <t2>rouge</t2> .`

Provides with this kind of log (a bit enriched by me comparing to the standard log):
```
   sentence = 0-Ceci 1-est 2-un 3-chapeau 4-rouge 5-.
   translation = 0-Here 1-is 2-a 3-red 4-hat 5-.
   alignment = 0-0 1-1 3-4 4-2 4-3 5-5
   suggestions = [
      ScoreEntry{memory=-1, language=fr -> en, sentence=[Ceci est un chapeau rouge .], translation=[Here is a red hat .], score=1.0}
   ]
```

With this result, where the tags are inverted:
`Here is <t2>a red</t2> <t1>hat</t1>.`

**Comment by nicolabertoldi** [comment]
Actually, this is the desired behaviour we implemented, and we think that is the correct way of projecting tags

I.e., tag `t1` surrounded `chapeau`in the source and it surrounds `hat`, which is the translation of `chapeau`.
Imagine that `t1` is a tag representing the bold character and `t2` is a tag representing the colorization of the words , you probably wants the the bold terms are `chapeau/hat` and that `rouge/red` are actually red-coloured.

**Comment by EtienneAb3d** [comment]
Yes, I know. It's the reason why I sent this request, with the experimental code, and the explanations above.

**Comment by nicolabertoldi** [comment]
@EtienneAb3d 

It is not clear to me why ModernMT approach could break the DTD when tags are "inverted".
Do you refer to cases where the DTD specifies explicitly the order the tags?
In this case, however there is no reason in my opinion to allow that the translation can  perform word-alignment.

We will take into account your solution, and discuss internally whether to include monotonic tag projection in ModernMT.

**Comment by EtienneAb3d** [comment]
It's more a question of principle : in a very safe process, it's more secure to ensure that all tags are strictly preserve in their original order, and also that the text is strictly placed in the same original tags (even if this is not the good texts in the good tags).

MS Word is often producing stupid redundant tag sequences, closing and opening always the same structure, without any kind of formatting change. In this tag structure, the text must always be strictly placed in tags defined as text containers. All text outside of these text containers is discarded when the file is opened at screen.

Here is a short example, suppose that `<t2>` is the tag that must contain the text:
Input:
`Ceci est un chapeau</t2></t1><t1><t2> rouge</t2></t1><t3/><t1><t2>.`
MMT output:
`Here is</t2></t1><t1><t2> a red</t2></t1> hat<t3/><t1><t2>.`

The text ` hat` is outside of a `<t2>` tag and will be discarded at screen. This is due to the fact that all `<t1><t2>` tags are moved to their best places according to the word alignments, letting the ` hat` text in a not well-defined in-between place.

My monotonic projection:
`Here is a red</t2></t1><t1><t2> hat</t2></t1><t3/><t1><t2>.`

Of course, the translated words is not strictly in the original tags, but in the case of stupid MS Word over-structuration, this has no effect. Now, no text is discarded at screen, and the file looks good.

**Comment by EtienneAb3d** [comment]
An other example (typical kind of inversion of the speaker in News sentences), suppose the `t2` and `t4` tags are the only ones that can contain text in the structure:
Input:
`<t1><t2>Le Premier Ministre</t2></t1><t3><t4> explique que tout ceci n'est pas compréhensible</t4></t3>`

MMT output:
`<t1><t2>All this is not understandable, Prime</t2></t1> Minister<t3> explain<t4></t4></t3>`

In this case, not only nearly all text is not in the good tags (in fact, only `Prime`), but also `Minister explain` is not in a text container and will be discarded at screen if this is a MS Word document.

My monotonic projection:
`<t1><t2>All</t2></t1><t3><t4> this is not understandable, Prime Minister explain</t4></t3>`

Of course, not perfect, but a larger part of the text is in the good tags (`this is not understandable` and `explain`), and no text is outside a text container. 

This is possible because of 2 complementary behaviors : 
1) the aligner tries to pair as much text as possible according to MMT word alignments (with a tolerant fuzzy positioning), and 
2) it tries to compress unpaired word sequences according to the direction where there should be matched according to MMT word alignments . 

In the pair trace below:
1) each word of the English part `All this is not understandable` is paired with one of its French MMT possible pairs
2) the French part `explique que tout` is compressed to the right, while the English part `Prime Minister explain` is compressed to the left, putting them all inside the `t4` tag enclosure
```
p0 <t1>(0=>0)~	----------	
p1 <t2>(0=>0)~	----------	
p2 Le(0)~[5]	----------	
p3 Premier(1)~[0]	----------	All(0)~[1]
p4 Ministre(2)~[6]	----------	
p5 </t2>(3=>1)~	----------	
p6 </t1>(3=>1)~	----------	
p7 <t3>(3=>1)~	----------	
p8 <t4>(3=>1)~	----------	
p9 explique(3)~[5][7]	----------	
p10 que(4)~[7]	----------	
p11 tout(5)~[4]	----------	
p12 ceci(6)~[1][4]	----------	this(1)~[6]
p13 n'(7)~[2][3]	----------	
p14 est(8)~[2]	----------	is(2)~[7][8]
p15 pas(9)~[3]	----------	not(3)~[7][9]
p16 compréhensible(10)~[4]	----------	understandable,(4)~[5][6][10]
p17 	----------	Prime(5)~[0][3]
p18 	----------	Minister(6)~[2]
p19 	----------	explain(7)~[3][4]
p20 </t4>(11=>8)~	----------	
p21 </t3>(11=>8)~	----------	

```
